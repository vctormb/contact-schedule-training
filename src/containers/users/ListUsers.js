
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export class ListUsers extends Component {
    state = {}

    componentDidMount() {
        this.fetchUsers();
    }

    fetchUsers = () => {
        const { dispatch } = this.props;

        dispatch({
            type: 'FETCH_USERS_REQUEST',
        });
    }

    editUser = (userId) => {
        const props = this.props;

        props.history.push(`/users/${userId}/edit`);
    }

    renderUsers = () => {
        const { usersReducer } = this.props;

        if (usersReducer.users.isLoading) {
            return <span className="fetching-users">Loading...</span>
        }

        return usersReducer.users.items.map((val, index) => {
            return (
                <div key={index}>
                    <span>{val.name} - </span>
                    <button className={`edit-user-${index}`} onClick={() => this.editUser(val.id)}>Edit</button>
                    <button className={`delete-user-${index}`} onClick={() => this.deleteUser(val.id)}>Delete</button>
                    <br />
                </div>
            )
        });
    }

    deleteUser = (userId) => {
        const { dispatch } = this.props;

        dispatch({
            type: 'DELETE_USER_REQUEST',
            payload: {
                userId
            }
        });
    }

    loadingDeleteUser = () => {
        return <span className="deleting-user">Loading...</span>
    }

    render() {
        const { usersReducer } = this.props;

        return (
            <div>
                {usersReducer.deleteUser.isLoading && this.loadingDeleteUser()}
                {this.renderUsers()}
                <Link to="/users/new" className="add-new-user">New user</Link>
            </div>
        );
    }
}

function mapStateToProps(state) {
    return {
        usersReducer: state.users
    }
}

export default connect(mapStateToProps)(ListUsers);