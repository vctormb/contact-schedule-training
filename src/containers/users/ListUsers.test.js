import React from 'react';
import PropTypes from 'prop-types';
import { shallow, mount } from 'enzyme';
import { ListUsers } from './ListUsers';
import reducer from '../../redux/reducers/users';

describe('<ListUsers />', () => {
    const dispatch = jest.fn();
    const push = jest.fn();

    let INITIAL_STATE = {
        usersReducer: {
            users: {
                items: [],
                isLoading: false,
            }
        }
    }

    let props = {
        dispatch,
        ...INITIAL_STATE,
        history: { push }
    };

    let listUsers = shallow(<ListUsers {...props} />);

    it('should render properly', () => {
        expect(listUsers).toMatchSnapshot();
    });

    describe('when componentDidMount()', () => {
        beforeEach(() => {
            listUsers = mount(<ListUsers {...props} />);
        });

        it('should dispatch FETCH_USERS_REQUEST action', () => {
            expect(dispatch).toBeCalledWith({type: "FETCH_USERS_REQUEST"});
        });
    });

    describe('when fetching users', () => {
        beforeEach(() => {
            INITIAL_STATE = {
                usersReducer: {
                    users: {
                        ...INITIAL_STATE.usersReducer.users,
                        isLoading: true
                    }
                }
            }

            props = { ...props, ...INITIAL_STATE }

            listUsers = shallow(<ListUsers {...props} />);
        });

        it('should render the loading message', () => {
            expect(listUsers.find('.fetching-users').exists()).toEqual(true);
        });
    });

    describe('when the list of users are not empty', () => {
        beforeEach(() => {
            INITIAL_STATE = {
                usersReducer: {
                    users: {
                        ...INITIAL_STATE.usersReducer.users,
                        items: [{ id: 1, name: 'Foo 1' }, { id: 2, name: 'Foo 2' }],
                        isLoading: false
                    }
                }
            }

            props = { ...props, ...INITIAL_STATE }

            listUsers = shallow(<ListUsers {...props} />);
        });

        it('should return the max number', () => {
            expect(listUsers.instance().props.usersReducer.users.items).toHaveLength(2);
        });

        describe('and the user wants to edit a user', () => {
            beforeEach(() => {
                push.mockClear();
            });

            it('should redirect to the edit page from user with id 1', () => {
                listUsers.find('.edit-user-0').simulate('click'); // first user
                expect(listUsers.instance().props.history.push).toHaveBeenCalledWith(`/users/${1}/edit`);
            });

            it('should redirect to the edit page from user with id 2', () => {
                listUsers.find('.edit-user-1').simulate('click'); // second user
                expect(listUsers.instance().props.history.push).toHaveBeenCalledWith(`/users/${2}/edit`);
            });
        });


    });
});