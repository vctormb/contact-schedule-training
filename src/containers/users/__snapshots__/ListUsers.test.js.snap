// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListUsers /> should render properly 1`] = `
<div>
  <Link
    className="add-new-user"
    replace={false}
    to="/users/new"
  >
    New user
  </Link>
</div>
`;
